rust/worker/src/system/types.rs

```diff
@@ -72,17 +74,96 @@ where
     }
 }
 
+/// A thin wrapper over a join handle that will panic if it is consumed twice.
+#[derive(Debug)]
+struct ConsumableJoinHandle {
+    handle: Option<tokio::task::JoinHandle<()>>,
+}
+
+impl ConsumableJoinHandle {
+    fn new(handle: tokio::task::JoinHandle<()>) -> Self {
+        ConsumableJoinHandle {
+            handle: Some(handle),
+        }
+    }
+
+    async fn consume(&mut self) -> Result<(), JoinError> {
+        match self.handle.take() {
+            Some(handle) => {
+                handle.await?;
+                Ok(())
+            }
+            None => {
+                panic!("Join handle already consumed");
+            }
+        }
+    }
+}
+
+/// A ComponentSender is generic over a component type. This struct is internal to the system module.
+/// It's implemented as a struct instead of a type alias so that it can implement common logic around the channel.
+///
+/// See ReceiverForMessage for a trait generic over a message type.
+#[derive(Debug)]
+pub(super) struct ComponentSender<C: Component> {
+    sender: tokio::sync::mpsc::Sender<WrappedMessage<C>>,
+}
+
+impl<C: Component> ComponentSender<C> {
+    pub(super) fn new(sender: tokio::sync::mpsc::Sender<WrappedMessage<C>>) -> Self {
+        ComponentSender { sender }
+    }
+
+    pub(super) async fn wrap_and_send<M>(
+        &self,
+        message: M,
+        tracing_context: Option<tracing::Span>,
+    ) -> Result<(), ChannelError>
+    where
+        C: Handler<M>,
+        M: Send + Debug + 'static,
+    {
+        self.sender
+            .send(wrap(message, tracing_context))
+            .await
+            .map_err(|_| ChannelError::SendError)
```

Comment:
was hoping to get rid of this and make it a type alias, but there's 3 separate places that would have to implement the same wrapping code here in that case

(and while these 4 lines are straightforward, the upcoming `.request()` method will be more complex and should really only be defined once)